### PR TITLE
boards: Jiminy-mega256rfr2: initial support

### PR DESCRIPTION
**Pull request by Josar**
The Jiminy board was designed by the IAS of the RWTH Aachen as a simple IoT board and features an atmega256rfr2. 

This commit adds the basic board files supporting xtimer, shell and led.

Later pull requests will add features for networking, RTC, PWM RGB LED, modules for the battery management and DC-DC converter and so on.

https://github.com/RIOT-OS/RIOT/pull/7738

@aabadie maybe this works.

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
@Josar thanks! This one looks way cleaner, though you still need so squash your commits. Please add the picture the RIOT wiki. This can be done by cloning it locally with  `git clone https://github.com/RIOT-OS/RIOT.wiki.git`. Add it to the `images` folder and push. If this doesn't work, just send it to me an I'll do it for you.

**Comment by Josar** [comment]
@PeterKietzmann  is it allright like that?

File: tests/pthread/Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 include ../Makefile.tests_common
 
-BOARD_BLACKLIST := arduino-mega2560 waspmote-pro arduino-uno arduino-duemilanove
-# arduino mega2560 uno duemilanove : unknown type name: clockid_t
```
Comment (aabadie) [inline tests/pthread/Makefile]: Please revert this change, here and in other places

File: cpu/atmega_common/periph/pm.c [hunk 2]
```diff
@@ -28,11 +30,20 @@
...
+    __asm__ __volatile__("mov r3, %0\n" :: "r" (0xAA));
+    #endif /* CPU_ATMEGA256RFR2 */
+
     /*
      * Since the AVR doesn't support a real software reset, we set the Watchdog
      * Timer on a 250ms timeout. Consider this a kludge.
      */
     irq_disable();
     wdt_enable(WDTO_250MS);
-    while(1);
```
Comment (aabadie) [inline cpu/atmega_common/periph/pm.c]: Looks unrelated, even if nearly valid change (missing space before `{}`)

File: cpu/atmega_common/periph/gpio.c [hunk 3]
```diff
@@ -290,8 +290,10 @@ ISR(INT6_vect, ISR_BLOCK)
 #endif
 
 #if defined(INT7_vect)
+#if !defined(BOARD_JIMINY_MEGA256RFR2) /**< INT7 is context swap */
```
Comment (aabadie) [inline cpu/atmega_common/periph/gpio.c]: could be done in the same `#if`: 
```c
#if defined(INT7_vect) && defined(BOARD_JIMINY_MEGA256RF2))
```

File: cpu/atmega256rfr2/startup.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * seems to be not replaceable without hacking inside the library. We
+ * circumvent the call to main by using section .init7 to call the function
+ * reset_handler which therefore is the real entry point and  section .init8
+ * which should never be reached but just in case jumps to exit.
+ * This way there should be no way to call main directly.
+ */
+void init7_ovr(void) __attribute__((section(".init7")));
+void init8_ovr(void) __attribute__((section(".init8")));
+
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/startup.c]: This blank line could be removed

File: cpu/atmega256rfr2/startup.c [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+#include <stdint.h>
+#include <avr/interrupt.h>
+#include <avr/io.h>
+
+/* For Catchall-Loop */
+#include "board.h"
+
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/startup.c]: This blank should be removed

File: cpu/atmega256rfr2/periph/cpuid.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+{
+
+    uint8_t  signature_0 = boot_signature_byte_get(0x00);
+    uint8_t  signature_1 = boot_signature_byte_get(0x02);
+    uint8_t  signature_2 = boot_signature_byte_get(0x04);
+
+    uint8_t  usr_sign_0 = boot_signature_byte_get(0x0100);
+    uint8_t  usr_sign_1 = boot_signature_byte_get(0x0102);
+
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/periph/cpuid.c]: This blank line should be removed

File: cpu/atmega256rfr2/periph/cpuid.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "periph/cpuid.h"
+
+#include "atmega_regs_common.h"
+#include "avr/boot.h"
+
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/periph/cpuid.c]: This blank line should be removed

File: cpu/atmega256rfr2/include/periph_cpu.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @name   Length of the CPU_ID in octets
+ * @{
+ */
+#define CPUID_LEN           (8U)
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/include/periph_cpu.h]: This blank line should be removed

File: cpu/atmega256rfr2/include/cpu_conf.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * @name Kernel configuration
+ *
+ * Since printf seems to get memory allocated by the linker/avr-libc the stack
+ * size tested sucessfully even with pretty small stacks.k
+ * @{
+ */
+
+/* keep
+ * THREAD_STACKSIZE_IDLE > THREAD_EXTRA_STACKSIZE_PRINTF
```
Comment (aabadie) [inline cpu/atmega256rfr2/include/cpu_conf.h]: put this on the same line above (no need to split)

File: cpu/atmega256rfr2/cpu.c [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+{
+    /* save the reset flags passed from the bootloader */
+    __asm__ __volatile__("mov %0, r2\n" : "=r" (mcusr_mirror) :);
+    __asm__ __volatile__("mov %0, r3\n" : "=r" (soft_rst) :);
+}
+
+
+void _reset_cause(void)
+{
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/cpu.c]: This blank line should be removed

File: cpu/atmega256rfr2/cpu.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+ * purposes.
+ * SCIRQS – Symbol Counter Interrupt Status Register
+ * BATMON – Battery Monitor Control and Status Register
+ * IRQ_STATUS /1 – Transceiver Interrupt Status Register
+ * EIFR – External Interrupt Flag Register
+ * PCIFR – Pin Change Interrupt Flag Register
+ */
+
+ISR(BADISR_vect){
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/cpu.c]: This blank line should be removed

File: cpu/atmega256rfr2/cpu.c [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    if(mcusr_mirror & (1 << PORF)) {
+        DEBUG("Power-on reset.\n");
+    }
+    if(mcusr_mirror & (1 << EXTRF)) {
+        DEBUG("External reset!\n");
+    }
+    if(mcusr_mirror & (1 << BORF)) {
+        DEBUG("Brownout reset!\n");
+    }
+    if(mcusr_mirror & (1 << WDRF)) {
```
Comment (aabadie) [inline cpu/atmega256rfr2/cpu.c]: missing space after `if` here and above and below

File: cpu/atmega256rfr2/cpu.c [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+uint8_t soft_rst __attribute__((section(".noinit")));
+void get_mcusr(void) __attribute__((naked)) __attribute__((section(".init0")));
+void get_mcusr(void)
+{
+    /* save the reset flags passed from the bootloader */
+    __asm__ __volatile__("mov %0, r2\n" : "=r" (mcusr_mirror) :);
+    __asm__ __volatile__("mov %0, r3\n" : "=r" (soft_rst) :);
+}
+
+
```
Comment (aabadie) [inline cpu/atmega256rfr2/cpu.c]: This blank line should be removed

File: boards/jiminy-mega256rfr2/include/periph_conf.h [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+ * @{
+ */
+#define SPI_NUMOF           1           /* set to 0 to disable SPI */
+/** @} */
+
+/**
+ * @name    I2C configuration
+ * @{
+ */
+#define I2C_NUMOF           1
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/include/periph_conf.h]: put parenthesis around value here and also for SPI_NUMOF, suffix with `U` as well

File: boards/jiminy-mega256rfr2/include/periph_conf.h [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+#include "periph_cpu_common.h"
+
+/**
+ * @name   Clock configuration
+ * @{
+ */
+#define CLOCK_CORECLOCK     (8000000UL)
+/** @} */
+
+
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/include/periph_conf.h]: This blank line should be removed

File: boards/jiminy-mega256rfr2/include/board.h [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+ * @{
+ */
+#define XTIMER_DEV          TIMER_DEV(0)
+#define XTIMER_CHAN         (0)
+#define XTIMER_WIDTH        (16)
+#define XTIMER_HZ           (125000UL)
+/** @} */
+
+/**
+ * @name Initialize board specific hardware, including clock, LEDs and std-IO
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/include/board.h]: Should be `@brief` here and also remove the doxygen block opening/closing, e.g. `@{` and `@}`

Comment (aabadie) [reply boards/jiminy-mega256rfr2/include/board.h]: just for board_init function

File: boards/jiminy-mega256rfr2/include/board.h [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+        DDRE |= (1 << PE7); \
+        EICRB |= (1 << ISC70); \
+        EIMSK |= (1 << INT7); \
+        sei(); \
+} while (0)
+#define AVR_CONTEXT_SWAP_INTERRUPT_VECT  INT7_vect
+#define AVR_CONTEXT_SWAP_TRIGGER   PORTE ^= (1 << PE7)
+/** @} */
+
+
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/include/board.h]: This blank line should be removed

File: boards/jiminy-mega256rfr2/Makefile.features [hunk 18]
```diff
@@ -0,0 +1,13 @@
+include $(RIOTBOARD)/common/arduino-atmega/Makefile.features
+
+# Put defined MCU peripherals here (in alphabetical order)
+
+
+
+# Various other features (if any)
+FEATURES_PROVIDED += periph_cpuid
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/Makefile.features]: Isn't this already provided by the CPU makefile ? If yes, you can remove this line

File: boards/jiminy-mega256rfr2/Makefile.features [hunk 19]
```diff
@@ -0,0 +1,13 @@
+include $(RIOTBOARD)/common/arduino-atmega/Makefile.features
+
+# Put defined MCU peripherals here (in alphabetical order)
+
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/Makefile.features]: No feature provided by the board ?

File: boards/jiminy-mega256rfr2/Makefile.features [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+
+
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/Makefile.features]: This blank line should be removed

**Comment by Josar** [comment]
@aabadie maybe now ;-)

And thanks for the great wok.

**Comment by PeterKietzmann** [comment]
@aabadie are you happy with the PR as is? I'll run some final tests now.

File: boards/jiminy-mega256rfr2/Makefile.include [hunk 21]
```diff
@@ -0,0 +1,26 @@
+# define the cpu used by the jiminy board
+export CPU = atmega256rfr2
+
+USEMODLUE += periph
```
Comment (aabadie) [inline boards/jiminy-mega256rfr2/Makefile.include]: typo on USEMODULE

Comment (aabadie) [reply boards/jiminy-mega256rfr2/Makefile.include]: by the way this line is also present in the CPU Makefile.include. is it really needed here ?

File: cpu/atmega256rfr2/doc.txt [hunk 22]
```diff
@@ -0,0 +1,10 @@
+/**
+ * @defgroup        cpu_atmega256rfr2 Atmel ATmega256rfr2
+ * @ingroup         cpu
+ * @brief           Implementation of Atmel's ATmega256rfr2 MCU
+ */
+
+/**
+ * @defgroup        cpu_atmega256rfr2_definitions Atmel ATmega256rfr2 Definitions
```
Comment (aabadie) [inline cpu/atmega256rfr2/doc.txt]: I don't think this group is needed

File: cpu/atmega256rfr2/periph/cpuid.c [hunk 23]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2018 RWTH Aachen, Josua Arndt, Steffen Robertz
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup  cpu_atmega256rfr2
```
Comment (aabadie) [inline cpu/atmega256rfr2/periph/cpuid.c]: should be `@ingroup` to be consistent with the other places

File: cpu/atmega256rfr2/doc.txt (hunk 22 above)
Comment (aabadie) [reply cpu/atmega256rfr2/doc.txt]: and it doesn't seem to be used actually

File: boards/jiminy-mega256rfr2/include/periph_conf.h [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+#define UART_1              MEGA_UART1
+#define UART_1_ISR          USART1_RX_vect
+/** @} */
+
+/**
+ * @name SPI configuration
+ *
+ * The atmega256rfr has only one hardware SPI with fixed pin configuration, so all
+ * we can do here, is to enable or disable it.
```
Comment (PeterKietzmann) [inline boards/jiminy-mega256rfr2/include/periph_conf.h]: Can you document the pin numbers here?

File: cpu/atmega_common/periph/pm.c [hunk 25]
```diff
@@ -28,6 +30,15 @@
 
 void pm_reboot(void)
 {
+    #if defined(CPU_ATMEGA256RFR2)
```
Comment (PeterKietzmann) [inline cpu/atmega_common/periph/pm.c]: Un-indent

File: tests/libfixmath_unittests/Makefile [hunk 26]
```diff
@@ -5,7 +5,8 @@ BOARD_BLACKLIST := arduino-mega2560 waspmote-pro arduino-uno arduino-duemilanove
 # differences in the toolchain)
 
 # The MSP boards don't feature round(), exp(), and log(), which are used in the unittests
-BOARD_BLACKLIST += chronos msb-430 msb-430h telosb wsn430-v1_3b wsn430-v1_4 z1
+BOARD_BLACKLIST += chronos msb-430 msb-430h telosb wsn430-v1_3b wsn430-v1_4 z1 \
+                   jiminy-mega256rfr2
```
Comment (PeterKietzmann) [inline tests/libfixmath_unittests/Makefile]: Remove blank line

File: tests/libfixmath_unittests/Makefile [hunk 27]
```diff
@@ -18,4 +19,4 @@ endif
 include $(RIOTBASE)/Makefile.include
 
 test:
-	tests/01-run.py
+    tests/01-run.py
```
Comment (PeterKietzmann) [inline tests/libfixmath_unittests/Makefile]: undo this change

File: boards/jiminy-mega256rfr2/include/periph_conf.h (hunk 24 above)
Comment (Josar) [reply boards/jiminy-mega256rfr2/include/periph_conf.h]: @PeterKietzmann done.

**Comment by Josar** [comment]
@PeterKietzmann 

SPI Test: 
short circuit the MISO and MOSI pins, labeled MOSI and MISO on the board ;-) 
Compile and flash test/periph_spi and then Type:
init 0 0 1 1 0
send What you like to type.

**Comment by PeterKietzmann** [comment]
@Josar thank. That's what I did be for and tried again just now. Still, I don't receive data. Need to check with a logic analyzer tomorrow.

**Comment by PeterKietzmann** [comment]
@Josar can you please check on your side if SPI still works? I don't see any signal neither on SCK nor on MOSI. I checked PB1 and PB2 pins, which are also labeled as SCK and MOSI on the board.

**Comment by PeterKietzmann** [comment]
@aabadie can you please check if your comments were addressed? @Josar Murdock fails, please fix related issues. However, builds look kind of good

File: tests/libfixmath_unittests/Makefile [hunk 28]
```diff
@@ -1,12 +1,12 @@
...
-BOARD_BLACKLIST := arduino-mega2560 waspmote-pro arduino-uno arduino-duemilanove
+BOARD_BLACKLIST := arduino-mega2560 waspmote-pro arduino-uno arduino-duemilanove \
+                   jiminy-mega256rfr2
 # arduino-mega2560: builds locally but breaks travis (possibly because of
 # differences in the toolchain)
-
 # The MSP boards don't feature round(), exp(), and log(), which are used in the unittests
 BOARD_BLACKLIST += chronos msb-430 msb-430h telosb wsn430-v1_3b wsn430-v1_4 z1
-
+
```
Comment (aabadie) [inline tests/libfixmath_unittests/Makefile]: trailing whitespaces

**Comment by Josar** [comment]
@PeterKietzmann for the murdock failure the adxx1234.h has to be fixed.
As INT1 is already defined in the iom256rfr2 header.

Any suggestions how to fix this properly?
I think this definition shouldn't be declared there anyway.
It is CPU Interrupt Pin Macro and should either be in the header of the cpu or in a RIOT/CPU/CPUx/include/missingStuffInManufacturerHeader.h


https://github.com/RIOT-OS/RIOT/blob/2cab4292d57cd6ec5d16752cfc60ee8030ed423d/drivers/include/adxl345.h#L95

**Comment by Josar** [comment]
SPI can be solved. Requires changes in this PR and a second PR for the test.

**Comment by Josar** [comment]
I allready fixed this SPI related stuff in the Jiminy related Part. Only have to commit it.

But It will require a second PR related to the test, which fixes the test as it does not call spi_init(). 

And you can not send something like `send something with white spaces`

As the arguments get splitted and the test only reads the first argument.

**Comment by PeterKietzmann** [comment]
Ah, now I see the problem. You need to call `periph_init()` from you `cpu_init` function as done [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/atmega328p/cpu.c#L29) for example. That function initializes the SPI bus once at startup (see [here](https://github.com/RIOT-OS/RIOT/blob/master/drivers/periph_common/init.c#L38)). Unfortunately I don't have the board with me to test right now.
As for the whitespaces I would assume putting quotes around the string should be fine?

**Comment by Josar** [comment]
I don't see why this should be done in `cpu_init` as the spi has only to be initialized when someone will actually use it.
And thats in this case the Test. IMHO.

Initializing everything in cpu_init is not what i understand as power aware!
I would rather activate all power save measures and force activation of each module when in use.

**Comment by PeterKietzmann** [comment]
@Josar please read the [SPI API description](http://riot-os.org/api/group__drivers__periph__spi.html) and [this thread](https://lists.riot-os.org/pipermail/devel/2017-May/005124.html).

**Comment by PeterKietzmann** [comment]
Besides the SPI discussion this PR can be rebased as #8696 has been merged

**Comment by Josar** [comment]
OK power save and reinitilization every aquire, still don't see why the pin config is not also in the aquire?

We set pins as output and then nobody uses them? Tri-State (High-Z) vs output low (sink)?
This somehow does not give me the right feeling.

I will do so for know but would be glad if there is a better way in future,

**Comment by PeterKietzmann** [comment]
> still don't see why the pin config is not also in the aquire?

The bus has always the same pin configuration. Imagine you have multiple slaves, you don't want to hand over the same configuration each time you aquire the bus. 

>I will do so for now but would be glad if there is a better way in future

To be honest, I doubt we'll find the manpower to change the SPI mechanism in RIOT again. But anyway, this lies beyond the scope of this PR. Maybe you wanna have a look at the proposal for a new I2C interface #6576?

>I cant find power_spi_enable(); where is it implemented

I think it's from the avr-libc. As I see it,  [atmega_common/include/atmega_regs_common.h](https://github.com/RIOT-OS/RIOT/blob/master/cpu/atmega_common/include/atmega_regs_common.h#L25) includes `avr/power.h`. So you don't have to implement it but rather include the common implementation.

As far as I know @roberthartung has a lot experience with avr based platforms as well as energy management in RIOT. If there are specific questions, I'm pretty sure he'd be happy to answer them.

@Josar I'm waiting for your "go" to start the next iteration of SPI testing

**Comment by Josar** [comment]
@PeterKietzmann OK should be good for testing now.

OK i just thought it through.
If someone does not wan't to use the pins for SPI he just exludes the MODULE everything is gone then.

But i am kind of confused by definition oft these:
```
FEATURES_REQUIRED
USEMODULE
```

Is there a docu? 
I didn't find it with a quick search. 
Is there a figure which lists where what is/or should be provided?
```
Application
board
common/board
cpu
common/cpu 
```

File: cpu/atmega256rfr2/cpu.c [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+    /* Set system clock Prescaler */
+    CLKPR = (1 << CLKPCE);  /* enable a change to CLKPR */
+    /* set the Division factor to 1 results in divisor 2 for internal Oscillator
+     * So FCPU = 8MHz
+     * */
+    CLKPR = 0;
+
+    /* Initialize peripherals, this is not a good solution */
+    /* SPI int, should not be done there*/
```
Comment (PeterKietzmann) [inline cpu/atmega256rfr2/cpu.c]: It makes absolutely no sense to comment the SPI concept here. In the state we merge this PR, it simply implements the RIOT architecture. If you have further ideas to improve this architecture, this should be discussed elsewhere.

**Comment by PeterKietzmann** [comment]
Tested SPI and I2C (with an external TSL2561 sensor) with success.

**Comment by PeterKietzmann** [comment]
Please fix Murdock error:
```
	cpu/atmega_common/periph/gpio.c:299: new blank line at EOF.
	cpu/atmega_common/periph/gpio.c:299: new blank line at EOF.
	tests/pthread_cleanup/Makefile:5: trailing whitespace.
	+# arduino mega2560 uno duemilanove : unknown type name: clockid_t                   
	ERROR: This change introduces new whitespace errors
```

**Comment by PeterKietzmann** [comment]
As a short answer to your question about `USEMODULE` vs `FEATURES_REQUIRED` see [this comment](https://github.com/RIOT-OS/RIOT/pull/8248#discussion_r156671985). If you have further questions to the documentation of the build system, could you please ask in #8479  or so?

File: cpu/atmega256rfr2/cpu.c (hunk 29 above)
Comment (Josar) [reply cpu/atmega256rfr2/cpu.c]: Sorrry this was a leftover, i fully agree with the RIOT architecture after thinking it through.

**Comment by Josar** [comment]
@PeterKietzmann hopefully i did not introduce new errors. ;-)

**Comment by PeterKietzmann** [comment]
@Josar we're not yet there, but really close :-). Commit messages may not increase 50 characters (see Murdock). Please fix and ping me again!

**Comment by Josar** [comment]
@PeterKietzmann  i am blind could not find it in murdock messages.
Anyhow reduced the commit messages and hope this works out now.

**Comment by PeterKietzmann** [comment]
One step further but still no there, yet. Please address:
```
	tests/pthread/Makefile:15: trailing whitespace.
	+	
	tests/pthread/Makefile:15: new blank line at EOF.
	tests/pthread_cleanup/Makefile:14: trailing whitespace.
	+	
	tests/pthread_cleanup/Makefile:14: new blank line at EOF.
	ERROR: This change introduces new whitespace errors
```

**Comment by Josar** [comment]
eclipse auto formatter!

**Comment by Josar** [comment]
@aabadie @PeterKietzmann  @shr70 thank you all for your patience and work.

**Comment by kYc0o** [comment]
Waaaao excellent! Congrats @Josar for the hardwork and all people involved. This opens the way for the porting of a board I have on my desk (RCB256RFR2-xpro) which uses the same CPU. I'll try to submit the PR soon.

Thanks!
